enclave-tls/src/tls_wrappers/openssl/un_negotiate.c

```diff
@@ -0,0 +1,279 @@
+/* Copyright (c) 2020-2021 Alibaba Cloud and Intel Corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define _GNU_SOURCE
+#include <string.h>
+#include <assert.h>
+#include <enclave-tls/log.h>
+#include <enclave-tls/err.h>
+#include <enclave-tls/tls_wrapper.h>
+#include "openssl.h"
+
+int etls_memcpy_s(void *dst, size_t dst_size, const void *src, size_t num_bytes)
+{
+	int result = -1;
+
+	if (dst == NULL)
+		ETLS_ERR("dst parameter is null pointer!\n");
+
+	if (src == NULL || dst_size < num_bytes) {
+		memset(dst, 0, dst_size);
+		ETLS_ERR("invalid parameters found!\n");
+	}
+
+	/* Reject overlapping buffers. */
+	if ((dst >= src && ((uint8_t *)dst < (uint8_t *)src + num_bytes)) ||
+	    (dst < src && ((uint8_t *)dst + dst_size > (uint8_t *)src))) {
+		memset(dst, 0, dst_size);
+		ETLS_ERR("there is overlapping copy here!\n");
+	}
+
+	memcpy(dst, src, num_bytes);
```

Comment:
so even num_bytes > dst_size and we still copy them and cause to a buffer overflow. So why we call it memcpy_s (secured)?

Reply:
This function is just a mess.

Reply:
have changed